crates/apub/src/http/mod.rs

```diff
@@ -52,6 +61,31 @@ pub async fn shared_inbox(
     .with_lemmy_type(FederationError::InboxTimeout.into())?
 }
 
+struct Dummy;
+
+impl ReceiveActivityHook<SharedInboxActivities, UserOrCommunity, LemmyContext> for Dummy {
+  async fn hook(
+    self,
+    activity: &SharedInboxActivities,
+    _actor: &UserOrCommunity,
+    context: &Data<LemmyContext>,
+  ) -> LemmyResult<()> {
+    // Store received activities in the database. This ensures that the same activity doesn't get
+    // received and processed more than once, which would be a waste of resources.
+    debug!("Received activity {}", activity.id().to_string());
+    ReceivedActivity::create(&mut context.pool(), &activity.id().clone().into()).await?;
+
+    // TODO: this could also take the actor as param, but lifetimes and serde derives are tricky
+    // TODO: this is really a before hook, but doesnt allow modifying the data. it could use a
+    // separate method so that error in plugin causes activity to be rejected.
+    plugin_hook_after("activity_received", activity)?;
+
+    // TODO: this method could also be used to check if actor is banned, instead of checking in each
+    // activity handler.
+    Ok(())
+  }
+}
```

Comment:
Main changes are here. There are many more possibilities especially for plugins, but they can be complicated to implement and its not clear if plugins will actually use them.